dask/dataframe/io/parquet/core.py

```diff
@@ -380,6 +380,7 @@ def to_parquet(
     compute=True,
     compute_kwargs=None,
     schema=None,
+    datafile_name_template="part.*.parquet",
```

Comment:
This is a rather long keyword argument name. Perhaps something like `filename_template=` (or something else) can be more concise while still conveying the same meaning.

On a related note, `to_csv` has a `name_function=` keyword argument for customizing the name of the output files which are saved. @martindurant do you have any thoughts on taking the same approach here to provide a bit more API consistency?

Reply:
Yes, I agree that would be nice - but there is a restriction here that the name can only be the basename of the path, because the rest will be generated. So it could be a template or a function to be passed an integer? That does make it tricky to infer any current max count number when appending. Perhaps when using this option, it's OK for the numbering to always start from zero, and its up to the user to provide a unique name?

Reply:
@jrbourbeau - thank you for commenting and for your great leadership on this project.  I am aligned with your goals for using good keyword argument names and for API consistency.  This is a public API change and we should be careful before merging it in!

The `to_csv` method gives 2 different ways to customize the filename and this PR introduces a third way, so I understand your hesitancy with this proposal.

Here's a few different examples of the current `name_function` functionality.

**Similar to the globstring syntax**

```python
pdf = pd.DataFrame(
    {"num1": [1, 2, 3, 4], "num2": [7, 8, 9, 10]},
)
df = dd.from_pandas(pdf, npartitions=2)
nice_filename = lambda x: f"hi-{x}.csv"
df.to_csv("./tmp/hello", name_function=nice_filename)
```

Outputted files:

```
hello/
  hi-0.csv.part
  hi-1.csv.part
```

(The `.part` looks like a little bug - I'll try to fix that in a separate PR).

**Combining with globstring**

```python
nice_filename = lambda x: f"hi-{x}.csv"
df.to_csv("./tmp/nice_filename2/ok-*", name_function=nice_filename)
```

```
nice_filename2/
  ok-hi-0.csv
  ok-hi-1.csv
```

**Write different UUIDs**

```python
my_fun = lambda x: f"{uuid.uuid4()}-{x}.csv"
df.to_csv("./tmp/uuid1", name_function=my_fun)
```

```
uuid1/
  87632d72-d617-4336-9a24-fe90aa195ce3-1.csv.part
  cab1aa55-d94f-44ea-86c0-ae251c7bd960-0.csv.part
```

**Write same UUID**

```python
id = uuid.uuid4()
my_fun = lambda x: f"{id}-{x}.csv"
df.to_csv("./tmp/uuid2", name_function=my_fun)
```

```
uuid2/
  a09433fd-3fa9-48a0-a212-19f699a3757a-0.csv.part
  a09433fd-3fa9-48a0-a212-19f699a3757a-1.csv.part
```

**My thoughts**

`name_function` seems more flexible than `filename_template` and is more consistent with `to_csv`.  It possible `name_function` with be harder for less technical users to understand.  Some folks have a difficult time with anonymous functions and passing functions as values.  I'm fine with either approach and will defer to your expertise.